File: lab/mar_21/solution.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <vector>

int main ()
{
  uint64_t n, m;
  std::cin >> n >> m;
  std::vector<std::vector<uint64_t>> friends(n + 1); // adjacency list
  while (m--)
  {
    uint64_t x, y;
    std::cin >> x >> y;
    friends[x].push_back(y);
    friends[y].push_back(x);
  }
  uint64_t max_sum {};
  for (uint64_t x {1}; x <= n; ++x)
  {
    auto is_friend_of_friend_of_x {std::vector<bool>(n + 1)};
    for (auto y : friends[x])
    {
      for (auto z : friends[y])
      {
        is_friend_of_friend_of_x[z] = (x != z);
      }
    }
    uint64_t sum {};
    for (uint64_t z {1}; z <= n; ++z)
    {
      if (is_friend_of_friend_of_x[z]) { sum += z; }
    }
    if (max_sum < sum) { max_sum = sum; }
  }
  std::cout << max_sum;
  return 0;
}

int main ()
{
  uint64_t n, m;
  std::cin >> n >> m;
  std::vector<std::vector<bool>> is_friend(n + 1, std::vector<bool>(n + 1)); // adjacency matrix
  while (m--)
  {
    uint64_t x, y;
    std::cin >> x >> y;
    is_friend[x][y] = is_friend[y][x] = true;
  }
  uint64_t max_sum {};
  for (uint64_t x {1}; x <= n; ++x)
  {
    auto is_friend_of_friend_of_x {std::vector<bool>(n + 1)};
    for (uint64_t y {1}; y <= n; ++y)
    {
      if (is_friend[x][y])
      {
        for (uint64_t z {1}; z <= n; ++z)
        {
          if (is_friend[y][z]) { is_friend_of_friend_of_x[z] = (x != z); }
        }
      }
    }
    uint64_t sum {};
    for (uint64_t z {1}; z <= n; ++z)
    {
      if (is_friend_of_friend_of_x[z]) { sum += z; }
    }
    if (max_sum < sum) { max_sum = sum; }
  }
  std::cout << max_sum;
  return 0;
}

/*
* author: redleaf
* email: redleaf23477@gapp.nthu.edu.tw
*/
#include <bits/stdc++.h>
using namespace std;
using LL = long long int;

int main() {
    int n, m; cin >> n >> m;
    // adjacency matrix
    vector<vector<bool>> graph(n, vector<bool>(n, false));
    for (int i = 0; i < m; i++) {
        int u, v; cin >> u >> v; u--, v--;
        graph[u][v] = graph[v][u] = true;
    }

    int ans = 0;

    for (int i = 0; i < n; i++) {
        vector<bool> vis(n, false);
        for (int u = 0; u < n; u++) if (graph[i][u]) {
            for (int v = 0; v < n; v++) if (graph[u][v]) {
                vis[v] = true;
            }
        }
        vis[i] = false;
        int sum = 0;
        for (int j = 0; j < n; j++) {
            if (vis[j]) sum += j+1;
        }
        ans = max(ans, sum);
    }

    cout << ans << "\n";
}

int main() {
    int n, m; cin >> n >> m;
    // adjacency list
    vector<vector<int>> graph(n);
    for (int i = 0; i < m; i++) {
        int u, v; cin >> u >> v; u--, v--;
        graph[u].emplace_back(v);
        graph[v].emplace_back(u);
    }

    int ans = 0;
    for (int i = 0; i < n; i++) {
        vector<bool> vis(n, false);
        for (auto u : graph[i]) {
            for (auto v : graph[u]) {
                vis[v] = true;
            }
        }
        int sum = 0;
        vis[i] = false;
        for (int j = 0; j < n; j++) {
            if (vis[j]) sum += j+1;
        }
        ans = max(sum, ans);
    }
    cout << ans << endl;
}

File: lab/mar_21/solution.py
#
# author: wang yang
# email: tnst92002@gmail.com
#
N = int(input())
M = int(input())

Adj = [[]for _ in range(N+1)] # adjacency list

for _ in range(M):
    x,y = map(int, input().split())
    Adj[x].append(y)
    Adj[y].append(x)

def Sol(Adj, v):
    hit = [0 for i in range(N+1)]
    for u in Adj[v]:
        for w in Adj[u]:
            hit[w] = w
    hit[v] = 0
    return sum(hit)

ans = 0
for i in range(1, N+1):
    ans = max(ans, Sol(Adj, i))
print(ans)

N = int(input())
M = int(input())

#
# author: wang yang
# email: tnst92002@gmail.com
#
Adj = [[0] * (N+1) for _ in range(N+1)] # adjacency matrix

for _ in range(M):
    x,y = map(int, input().split())
    Adj[x][y] = Adj[y][x] = 1

def Sol(Adj, v):
    hit = [0 for i in range(N+1)]
    for u in range(1, N+1):
        if Adj[v][u]:
            for w in range(1, N+1):
                if Adj[u][w]:
                    hit[w] = w
    hit[v] = 0
    return sum(hit)

ans = 0
for i in range(1, N+1):
    ans = max(ans, Sol(Adj, i))
print(ans)

#
# author: redleaf
# email: redleaf23477@gapp.nthu.edu.tw
#
def adj_list():
    n = int(input())
    m = int(input())
    graph = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input().split())
        u, v = u-1, v-1
        graph[u].append(v)
        graph[v].append(u)

    ans = 0
    for i in range(n):
        vis = [0] * n
        for u in graph[i]:
            for v in graph[u]:
                vis[v] = 1
        vis[i] = 0
        ans = max(ans, sum((i+1) * vis[i] for i in range(n)))
    print(ans)

def adj_mat():
    n = int(input())
    m = int(input())
    graph = [[0] * n for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input().split())
        u, v = u-1, v-1
        graph[u][v] = graph[v][u] = 1

    ans = 0
    for i in range(n):
        vis = [0] * n
        for u in [u for u in range(n) if graph[i][u] == 1]:
            for v in [v for v in range(n) if graph[u][v] == 1]:
                vis[v] = 1
        vis[i] = 0
        ans = max(ans, sum((i+1) * vis[i] for i in range(n)))
    print(ans)

if __name__=='__main__':
    adj_mat()